File: docs/submit/bear/python/fsm.py
# fsm.py  ── Python port of the engine.
# Same week tags as fsm.lua. Uncomment one block per week.
# ─────────────────────────────────────────────────────
import sys, os, random, time

# S1 ── seed ──────────────────────────────────────────
SEED = int(sys.argv[1]) if len(sys.argv)>1 else \
       int(os.getenv("SEED","0")) or int(time.time())
random.seed(SEED)
print(f"-- seed {SEED}  (replay: python bear.py {SEED})")

# S1 ── id counter ────────────────────────────────────
_id = 0
def new_id():
    global _id; _id += 1; return _id

# S1 ── shuffle ───────────────────────────────────────
def shuffle(lst):
    import random as r
    lst = list(lst); r.shuffle(lst); return lst

# S2 ── token registry ────────────────────────────────
# from dataclasses import dataclass, field
# from enum import Enum
# _tokens: dict[str, "Token"] = {}
#
# @dataclass
# class Token:
#     name: str
#     id:   int
#     def __repr__(self): return f"Token({self.name})"
#
# def make(name: str) -> Token:
#     if name in _tokens:
#         print(f"WARNING: duplicate token '{name}'")
#         return _tokens[name]
#     t = Token(name=name, id=new_id())
#     _tokens[name] = t
#     return t

# S3 ── comparators ───────────────────────────────────
# from functools import partial
# import operator as op
#
# def _cmp(fn, token, n):
#     return lambda p: fn(getattr(p, token.name, 0), n)
#
# def over(token, n): return _cmp(op.gt,  token, n)
# def to  (token, n): return _cmp(op.le,  token, n)
# def eq  (token, n): return _cmp(op.eq,  token, n)

# S4 ── say() ─────────────────────────────────────────
# def say(msg: str):
#     def _action(p):
#         def replacer(k):
#             v = getattr(p, k, None)
#             if v is None:
#                 print(f"WARNING: unknown key '{k}' in say()")
#                 return f"<{k}?>"
#             return str(v)
#         import re
#         print(re.sub(r"\{(\w+)\}",
#               lambda m: replacer(m.group(1)), msg))
#     return _action

# S5 ── now() ─────────────────────────────────────────
# def now(token, delta=None):
#     def _update(p):
#         if delta is None:
#             setattr(p, token.name, True); return True
#         if delta == 0:
#             setattr(p, token.name, False); return False
#         v = getattr(p, token.name, 0) + delta
#         setattr(p, token.name, v); return v
#     return _update

# S6 ── eval_when(): or-of-ands ───────────────────────
# def eval_when(clauses, p):
#     for conjunct in clauses:
#         preds = conjunct.get("preds", [])
#         if all(pred(p) for pred in preds):
#             return conjunct.get("target")
#     return None

# S7 ── lint() ────────────────────────────────────────
# def lint(agent):
#     rules, start = agent.states, agent.start
#     reachable = set()
#     for sname, state in rules.items():
#         for tgt in (state.get("transitions") or {}).values():
#             reachable.add(tgt)
#             if tgt not in rules:
#                 print(f"LINT [{agent.name}]: ghost state "
#                       f"'{tgt}' from '{sname}'")
#     for sname, state in rules.items():
#         t = state.get("transitions") or {}
#         if len(t)==0 and not state.get("terminal"):
#             print(f"LINT [{agent.name}]: dead end '{sname}'")
#     for sname in rules:
#         if sname != start and sname not in reachable:
#             print(f"LINT [{agent.name}]: "
#                   f"unreachable state '{sname}'")

# S1 ── core run (iterative; Python has no TCO) ───────
def _run(agent):
    rules = agent.states
    while True:
        s     = agent.current
        state = rules.get(s)
        if not state: return agent

        # S1: action
        action = state.get("action")
        if action: action(agent)

        # S5: update  [uncomment S5]
        # update = state.get("update")
        # if update: update(agent)

        # S6: exit    [uncomment S6]
        # exit_fn = state.get("exit")

        if not agent.queue: return agent
        e = agent.queue.pop(0)

        # S6: guard override  [uncomment S6]
        # guard_target = eval_when(state.get("guard") or [], agent)

        # S10: wildcard  [uncomment S10]
        # target = ((state.get("transitions") or {}).get(e)
        #        or (state.get("transitions") or {}).get("*")
        #        or s)
        target = (state.get("transitions") or {}).get(e, s)

        # S10: trace  [uncomment S10]
        # agent.trace.append(f"[{e}] {s} -> {target}")

        # S6: fire exit before move  [uncomment S6]
        # if exit_fn: exit_fn(s, agent)

        agent.current = target

def start(agent):
    agent.current = agent.current or agent.start
    return _run(agent)

# S8 ── step(): one event per tick  [uncomment S8] ────
# def step(agent):
#     if not agent.queue: return
#     _run_one(agent)   # variant of _run that pops one event

# S10 ── print_trace()  [uncomment S10] ───────────────
# def print_trace(agent):
#     for line in getattr(agent, "trace", []):
#         print("  " + line)

File: docs/submit/bear/lua/fsm.lua
-- fsm.lua  ── the engine.  uncomment one block per week.
-- never rewrite; only reveal.
-- ─────────────────────────────────────────────────────

-- S1 ── seed + random ─────────────────────────────────
local SEED = tonumber(arg[1])
         or  tonumber(os.getenv("SEED"))
         or  os.time()
math.randomseed(SEED)
io.write(string.format("-- seed %d  (replay: lua bear.lua %d)\n",
  SEED, SEED))

-- S1 ── shuffle (Fisher-Yates) ────────────────────────
local function shuffle(t)
  for i = #t, 2, -1 do
    local j = math.random(i)
    t[i], t[j] = t[j], t[i] end
  return t end

-- S1 ── global id counter ─────────────────────────────
local _id = 0
local function new_id() _id = _id + 1; return _id end

-- S2 ── token registry ────────────────────────────────
-- local _tokens = {}
-- local function make(name)
--   if _tokens[name] then
--     io.write("WARNING: duplicate token '"..name.."'\n")
--     return _tokens[name] end
--   local t = {name=name, id=new_id()}
--   _tokens[name] = t
--   return t end

-- S4 ── say(): interpolating action strings ───────────
-- local function say(msg)
--   return function(p)
--     local s = msg:gsub("{(%w+)}", function(k)
--       local v = p[k]
--       if v == nil then
--         io.write("WARNING: unknown key '"..k.."' in say()\n")
--         return "<"..k.."?>" end
--       return tostring(v) end)
--     print(s) end end

-- S5 ── now(): mutate payload tokens ──────────────────
-- local function now(token, delta)
--   return function(p)
--     if delta == nil then p[token.name]=true; return true end
--     if delta == 0   then p[token.name]=false; return false end
--     p[token.name] = (p[token.name] or 0) + delta
--     return p[token.name] end end

-- S3 ── comparators (pure, no side-effects) ───────────
-- local function over(token, n)
--   return function(p) return (p[token.name] or 0) > n end end
-- local function to(token, n)
--   return function(p) return (p[token.name] or 0) <= n end end
-- local function eq(token, n)
--   return function(p) return p[token.name] == n end end

-- S6 ── eval_when(): or-of-ands guard evaluation ──────
-- local function eval_when(clauses, p)
--   for _, conjunct in ipairs(clauses) do
--     local ok = true
--     for _, pred in ipairs(conjunct) do
--       if not pred(p) then ok=false; break end end
--     if ok then return conjunct.target end end
--   return nil end

-- S7 ── lint(): static analysis ───────────────────────
-- local function lint(agent)
--   local rules, start = agent.states, agent.start
--   local reachable = {}
--   -- ghost states + build reachable set
--   for sname, state in pairs(rules) do
--     for _, tgt in pairs(state.transitions or {}) do
--       reachable[tgt] = true
--       if not rules[tgt] then
--         io.write(string.format(
--           "LINT [%s]: ghost state '%s' from '%s'\n",
--           agent.name, tgt, sname)) end end end
--   -- dead ends
--   for sname, state in pairs(rules) do
--     local t = state.transitions or {}
--     local n = 0
--     for _ in pairs(t) do n=n+1 end
--     if n==0 and not state.terminal then
--       io.write(string.format(
--         "LINT [%s]: dead end '%s'\n", agent.name, sname)) end end
--   -- unreachable
--   for sname in pairs(rules) do
--     if sname ~= start and not reachable[sname] then
--       io.write(string.format(
--         "LINT [%s]: unreachable state '%s'\n",
--         agent.name, sname)) end end end

-- S1 ── core run (TCO) ────────────────────────────────
local function run(rules, s, p)
  local state = rules[s]
  if not state then return p end

  -- S1: action always fires
  if state.action then state.action(p) end

  -- S5: update/now mutations  [uncomment S5]
  -- if state.update then state.update(p) end

  -- S6: guard selects next state  [uncomment S6]
  -- (used in transition lookup below)

  -- S6: exit fires before leaving  [uncomment S6]
  -- local function depart(next_s)
  --   if state.exit then state.exit(s, p) end
  --   return next_s end

  local e = table.remove(p.queue, 1)
  if not e then return p end

  -- S10: wildcard fallback  [uncomment S10]
  -- local target = (state.transitions or {})[e]
  --             or (state.transitions or {})["*"]
  --             or s
  local target = (state.transitions or {})[e] or s  -- S1 version

  -- S10: trace log  [uncomment S10]
  -- p.trace = p.trace or {}
  -- p.trace[#p.trace+1] =
  --   string.format("[%s] %s -> %s", e, s, target)

  return run(rules, target, p) end  -- TCO

-- S8 ── step(): one event per tick ────────────────────
-- local function step(agent)
--   if #agent.queue == 0 then return end
--   run(agent.states, agent.current, agent) end

-- S1 ── start(): run to completion ────────────────────
local function start(agent)
  agent.current = agent.current or agent.start
  return run(agent.states, agent.current, agent) end

-- S10 ── print_trace() ────────────────────────────────
-- local function print_trace(agent)
--   for _, line in ipairs(agent.trace or {}) do
--     print("  "..line) end end

return {
  new_id      = new_id,
  start       = start,
  shuffle     = shuffle,
  -- S2: make   = make,
  -- S3: over   = over, to=to, eq=eq,
  -- S4: say    = say,
  -- S5: now    = now,
  -- S6: eval_when = eval_when,
  -- S7: lint   = lint,
  -- S8: step   = step,
  -- S10: print_trace = print_trace,
}
